grpc/errors.go

```diff
@@ -63,59 +72,83 @@ func unwrapError(err error, md metadata.MD) error {
 		return nil
 	}
 
-	unwrappedErr := status.Convert(err).Message()
-
 	errTypeStrs, ok := md["errortype"]
 	if !ok {
 		return err
 	}
+
+	inErrMsg := status.Convert(err).Message()
 	if len(errTypeStrs) != 1 {
 		return berrors.InternalServerError(
-			"multiple errorType metadata, wrapped error %q",
-			unwrappedErr,
+			0,
+			"multiple 'errortype' metadata, wrapped error %q",
+			inErrMsg,
 		)
 	}
 
-	errType, decErr := strconv.Atoi(errTypeStrs[0])
+	inErrType, decErr := strconv.Atoi(errTypeStrs[0])
 	if decErr != nil {
 		return berrors.InternalServerError(
+			0,
 			"failed to decode error type, decoding error %q, wrapped error %q",
 			decErr,
-			unwrappedErr,
+			inErrMsg,
 		)
 	}
-	outErr := berrors.New(berrors.ErrorType(errType), unwrappedErr)
-
-	subErrsJSON, ok := md["suberrors"]
-	if !ok {
-		return outErr
-	}
-	if len(subErrsJSON) != 1 {
-		return berrors.InternalServerError(
-			"multiple suberrors metadata, wrapped error %q",
-			unwrappedErr,
+	inErr := berrors.New(berrors.ErrorType(inErrType), 0, inErrMsg)
+	var outErr *berrors.BoulderError
+	if !errors.As(inErr, &outErr) {
```

Comment:
I realize that this is copy-pasted from the existing code, but doesn't this result in a double-indirection? The second argument to errors.As is a pointer to a pointer to a berrors.BoulderError.

Reply:
It does, but the double-indirection is correct. The content of `inErr` is of type `*berrors.BoulderError`. `errors.As` wants a pointer it can write to (if the unwrapping matches). That means a pointer to a `*berrors.BoulderError`, or a `**berrors.BoulderError`.

Reply:
I guess my main question was "but why the type of inErr/outErr `*berrors.BoulderError`?", and I guess the answer is "because that's what `berrors.New` returns". Even though it's obviously conventional for `.New()` to return a reference to something allocated on the heap, it didn't even occur to me that it's `*berrors.BoulderError` that implements the `error` interface, not `berrors.BoulderError`.